{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os.path\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from scipy import linalg\n",
    "from scipy.sparse import csr_matrix\n",
    "from sklearn.preprocessing import OneHotEncoder\n",
    "\n",
    "import Bio.PDB, warnings\n",
    "pdb_list = Bio.PDB.PDBList()\n",
    "pdb_parser = Bio.PDB.PDBParser()\n",
    "from scipy.spatial import distance_matrix\n",
    "from Bio import BiopythonWarning\n",
    "warnings.simplefilter('ignore', BiopythonWarning)\n",
    "\n",
    "from joblib import Parallel, delayed\n",
    "import timeit\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "%matplotlib inline\n",
    "\n",
    "# # --- Import our Code ---# #\n",
    "#import emachine as EM\n",
    "from direct_info import direct_info\n",
    "\n",
    "# import data processing and general DCA_ER tools\n",
    "from data_processing import data_processing\n",
    "import ecc_tools as tools\n",
    "from pathlib import Path\n",
    "np.random.seed(1)\n",
    "\n",
    "\n",
    "def check_symmetric(a, rtol=1e-05, atol=1e-08):\n",
    "    return np.allclose(a, a.T, rtol=rtol, atol=atol)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Read in Protein Structure\n",
    "1. We will start by reading in a set of Protein structures with the given Protein ID \"pfam_id\"\n",
    "    * [More Info on working with PDB structures in DataFrames](http://rasbt.github.io/biopandas/tutorials/Working_with_PDB_Structures_in_DataFrames/)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "number of pdb structures: 2\n"
     ]
    }
   ],
   "source": [
    "data_path = Path('/home', 'eclay','Pfam-A.full')\n",
    "data_path = Path('Pfam-A.full')\n",
    "DCA_ER_dir = Path('/home/evan/PycharmProjects/DCA_ER/')\n",
    "\n",
    "\n",
    "\n",
    "pfam_id = 'PF00023'\n",
    "pfam_id = 'PF00011'\n",
    "pfam_id = 'PF00186'\n",
    "pfam_id = 'PF14525'\n",
    "pfam_id = 'PF03496'\n",
    "\n",
    "\n",
    "n_jobs = 6\n",
    "create_new = True\n",
    "\n",
    "\n",
    "if 0:\n",
    "    DCA_ER_dir = '/home/eclay/DCA_ER/'\n",
    "    msa_npy_file = '/home/eclay/Pfam-A.full/%s/msa.npy' % pfam_id # Hurricane Location\n",
    "    msa_fa_file  = '/home/eclay/Pfam-A.full/%s/msa.fa' % pfam_id # Hurricane Location\n",
    "    pdb_ref_file = '/home/eclay/Pfam-A.full/%s/pdb_refs.npy' % pfam_id # Hurricane Location\n",
    "if 1:\n",
    "    DCA_ER_dir = '/home/evan/PycharmProjects/DCA_ER/'\n",
    "    msa_npy_file = '/home/evan/PycharmProjects/DCA_ER/Pfam-A.full/%s/msa.npy' % pfam_id\n",
    "    msa_fa_file  = '/home/evan/PycharmProjects/DCA_ER/Pfam-A.full/%s/msa.fa' % pfam_id\n",
    "    pdb_ref_file = '/home/evan/PycharmProjects/DCA_ER/Pfam-A.full/%s/pdb_refs.npy' % pfam_id\n",
    "    out_dir = '%sprotein_data/di/' % DCA_ER_dir\n",
    "if 0:\n",
    "    DCA_ER_dir = '/home/ecresswell/DCA_ER/'\n",
    "    msa_npy_file = '/home/ecresswell/DCA_ER/Pfam-A.full/%s/msa.npy' % pfam_id\n",
    "    msa_fa_file  = '/home/ecresswell/DCA_ER/Pfam-A.full/%s/msa.fa' % pfam_id\n",
    "    pdb_ref_file = '/home/ecresswell/DCA_ER/Pfam-A.full/%s/pdb_refs.npy' % pfam_id\n",
    "    out_dir = '%sprotein_data/di/' % DCA_ER_dir\n",
    "if 1:\n",
    "    DCA_ER_dir = '/data/cresswellclayec/DCA_ER/'\n",
    "    msa_npy_file = '/data/cresswellclayec/DCA_ER/Pfam-A.full/%s/msa.npy' % pfam_id\n",
    "    msa_fa_file  = '/data/cresswellclayec/DCA_ER/Pfam-A.full/%s/msa.fa' % pfam_id\n",
    "    pdb_ref_file = '/data/cresswellclayec/DCA_ER/Pfam-A.full/%s/pdb_refs.npy' % pfam_id\n",
    "    out_dir = '%sprotein_data/di/' % DCA_ER_dir\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# Set DCA_ER directory\n",
    "DCA_dir = os.getcwd()\n",
    "\n",
    "# Define data directories\n",
    "# Need to think on best way to do this..\n",
    "# Referencing the same dataframe may be useful so we dont always have to load individual ref files...\n",
    "# however we also\n",
    "individual_pdb_ref_file = Path(data_path, pfam_id, 'pdb_refs.npy')\n",
    "pdb = np.load(individual_pdb_ref_file)\n",
    "processed_data_dir = \"%s/protein_data/data_processing_output\" % DCA_dir\n",
    "\n",
    "# delete 'b' in front of letters (python 2 --> python 3)\n",
    "pdb = np.array([pdb[t,i].decode('UTF-8') for t in range(pdb.shape[0]) \\\n",
    "         for i in range(pdb.shape[1])]).reshape(pdb.shape[0],pdb.shape[1])\n",
    "\n",
    "\n",
    "# Print number of pdb structures in Protein ID folder\n",
    "npdb = pdb.shape[0]\n",
    "print('number of pdb structures:',npdb)\n",
    "\n",
    "# Create pandas dataframe for protein structure\n",
    "pdb_df = pd.DataFrame(pdb,columns = ['PF','seq','id','uniprot_start','uniprot_end',\\\n",
    "                                 'pdb_id','chain','pdb_start','pdb_end'])\n",
    "pdb_df.head()\n",
    "\n",
    "# print(\"Direct Information from Expectation reflection:\\n\",di)\n",
    "def no_diag(mat, diag_l, s_index=None, make_big=False):\n",
    "    rows, columns = mat.shape\n",
    "    if make_big:\n",
    "        new_mat = 100. * np.ones((rows,columns))\n",
    "    else:\n",
    "        new_mat = np.zeros((rows,columns))\n",
    "    for row in range(rows):\n",
    "        for col in range(columns):\n",
    "            if s_index is None:\n",
    "                if abs(row-col) > diag_l:\n",
    "                    new_mat[row, col] = mat[row ,col]\n",
    "            else:\n",
    "                if abs(s_index[row]-s_index[col]) > diag_l:\n",
    "                    new_mat[row, col] = mat[row ,col]    \n",
    "    return new_mat\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "PDB DataFrame needs to have PF######, PDB-id, as well as start and end locations of amino acids, as well as reference sequence in msa (not in current file with all pdb data)\n",
    "### PDB Website Search Engine\n",
    "   * We can visualize the complex structure of the proteins\n",
    "   * The data-base below gives further information on the proteins above (input pdb_id)\n",
    "\n",
    "### Define contact map\n",
    "2. Import contact map functions from newly defined tools python module"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import ecc_tools as tools"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "seq: 445\n",
      "Original Sequence length:  912\n",
      "pdb ref example (pdb[0])  (after UTF-8 decode, removing 'b'):\n",
      " ['PF03496' '445' 'Q88A91_PSESM' '108' '258' '3U0J' 'B' '108' '258']\n",
      "tpdb (s_ipdb) is :  445\n",
      "#\n",
      "\n",
      "-------------------------Remove Gaps--------------------------#\n",
      "Shape of s is :  (469, 912)\n",
      "s = \n",
      " [['-' '-' '-' ... '-' '-' '-']\n",
      " ['-' '-' '-' ... '-' '-' '-']\n",
      " ['-' '-' '-' ... '-' '-' '-']\n",
      " ...\n",
      " ['-' '-' '-' ... '-' '-' '-']\n",
      " ['-' '-' '-' ... '-' '-' '-']\n",
      " ['-' '-' '-' ... '-' '-' '-']]\n",
      "s[tpdb] shape is  (151,)\n",
      "s = \n",
      " [['s' 't' 'l' ... 'i' 's' 'l']\n",
      " ['-' '-' '-' ... '-' '-' '-']\n",
      " ['m' 'v' 'i' ... 'g' 'v' 'i']\n",
      " ...\n",
      " ['i' 'n' 'n' ... 'd' 'g' 'm']\n",
      " ['h' 'g' 'l' ... 'p' 'd' 'p']\n",
      " ['t' 'y' 'i' ... 'k' 't' 'l']]\n",
      "though s still has gaps, s[445] does not:\n",
      " ['t' 'p' 'l' 'Y' 'R' 'E' 'V' 'N' 'N' 'Y' 'L' 'r' 'l' 'q' 'H' 'E' 'N' 'S'\n",
      " 'G' 'R' 'E' 'A' 'E' 'I' 'D' 'N' 'h' 'd' 'E' 'K' 'L' 'S' 'P' 'H' 'I' 'K'\n",
      " 'M' 'L' 'S' 'S' 'A' 'L' 'N' 'R' 'L' 'm' 'd' 'V' 'A' 'A' 'F' 'R' 'G' 'T'\n",
      " 'V' 'Y' 'R' 'G' 'I' 'R' 'G' 'D' 'L' 'D' 'T' 'I' 'A' 'R' 'L' 'Y' 'H' 'L'\n",
      " 'F' 'D' 'T' 'G' 'G' 'R' 'Y' 'V' 'E' 'P' 'A' 'F' 'M' 'S' 'T' 'T' 'R' 'I'\n",
      " 'K' 'D' 'S' 'A' 'Q' 'V' 'F' 'E' 'P' 'G' 't' 'p' 'n' 'N' 'I' 'A' 'F' 'Q'\n",
      " 'I' 'S' 'L' 'K' 'R' 'G' 'A' 'D' 'I' 'S' 'G' 'S' 'S' 'Q' 'A' 'P' 'S' 'E'\n",
      " 'E' 'E' 'I' 'M' 'L' 'P' 'M' 'M' 'S' 'E' 'F' 'V' 'I' 'E' 'H' 'A' 'S' 'A'\n",
      " 'L' 'S' 'e' 'g' 'k' 'h' 'l']\n",
      "s shape is  (469, 151)\n",
      "Saving indices of reference sequence s[445](length=151):\n",
      " [  0   1   2   3   4   5   6   7   8   9  10  11  12  13  14  15  16  17\n",
      "  18  19  20  21  22  23  24  25  26  27  28  29  30  31  32  33  34  35\n",
      "  36  37  38  39  40  41  42  43  44  45  46  47  48  49  50  51  52  53\n",
      "  54  55  56  57  58  59  60  61  62  63  64  65  66  67  68  69  70  71\n",
      "  72  73  74  75  76  77  78  79  80  81  82  83  84  85  86  87  88  89\n",
      "  90  91  92  93  94  95  96  97  98  99 100 101 102 103 104 105 106 107\n",
      " 108 109 110 111 112 113 114 115 116 117 118 119 120 121 122 123 124 125\n",
      " 126 127 128 129 130 131 132 133 134 135 136 137 138 139 140 141 142 143\n",
      " 144 145 146 147 148 149 150]\n",
      "#--------------------------------------------------------------#\n",
      "\n",
      "\n",
      "removing non aligned (lower case) columns in subject sequence:\n",
      "  [  0   1   2  11  12  13  26  27  45  46 100 101 102 146 147 148 149 150] \n",
      "\n",
      "found 3 duplicates! (Removing...)\n",
      "After removing bad sequences, tpdb is now  442\n",
      "(466, 151)\n",
      "In Data Processing Reference Sequence (shape= (151,) ): \n",
      " ['t' 'p' 'l' 'Y' 'R' 'E' 'V' 'N' 'N' 'Y' 'L' 'r' 'l' 'q' 'H' 'E' 'N' 'S'\n",
      " 'G' 'R' 'E' 'A' 'E' 'I' 'D' 'N' 'h' 'd' 'E' 'K' 'L' 'S' 'P' 'H' 'I' 'K'\n",
      " 'M' 'L' 'S' 'S' 'A' 'L' 'N' 'R' 'L' 'm' 'd' 'V' 'A' 'A' 'F' 'R' 'G' 'T'\n",
      " 'V' 'Y' 'R' 'G' 'I' 'R' 'G' 'D' 'L' 'D' 'T' 'I' 'A' 'R' 'L' 'Y' 'H' 'L'\n",
      " 'F' 'D' 'T' 'G' 'G' 'R' 'Y' 'V' 'E' 'P' 'A' 'F' 'M' 'S' 'T' 'T' 'R' 'I'\n",
      " 'K' 'D' 'S' 'A' 'Q' 'V' 'F' 'E' 'P' 'G' 't' 'p' 'n' 'N' 'I' 'A' 'F' 'Q'\n",
      " 'I' 'S' 'L' 'K' 'R' 'G' 'A' 'D' 'I' 'S' 'G' 'S' 'S' 'Q' 'A' 'P' 'S' 'E'\n",
      " 'E' 'E' 'I' 'M' 'L' 'P' 'M' 'M' 'S' 'E' 'F' 'V' 'I' 'E' 'H' 'A' 'S' 'A'\n",
      " 'L' 'S' 'e' 'g' 'k' 'h' 'l']\n",
      "99\n",
      "After removing bad sequences, tpdb is now  348\n",
      "\n",
      "After removing bad sequences...\n",
      "tpdb (s_ipdb) is :  348\n",
      "(367, 151)\n",
      "found bad columns := [  0  11  12  13  18  19  20  21  22  24  25  26  27  45  46  59  60  61\n",
      "  62  63  71  92  93 100 101 102 103 142 143 144 145]\n",
      "found conserved columns (80% repetition):\n",
      " [  7  11  12  13  26  45  56  85 127]\n",
      "We remove conserved and bad columns with, at the following indices:\n",
      " [  0   1   2   7  11  12  13 142 143 144 145  18  19  20  21  22 146  24\n",
      "  25  26  27 147 148 149 150  45  46  56  59  60  61  62  63  71  85  92\n",
      "  93 100 101 102 103 127]\n",
      "Removed Columns...\n",
      "s now has shape:  (367, 109)\n",
      "s_index (length=109) = \n",
      " [  3   4   5   6   8   9  10  14  15  16  17  23  28  29  30  31  32  33\n",
      "  34  35  36  37  38  39  40  41  42  43  44  47  48  49  50  51  52  53\n",
      "  54  55  57  58  64  65  66  67  68  69  70  72  73  74  75  76  77  78\n",
      "  79  80  81  82  83  84  86  87  88  89  90  91  94  95  96  97  98  99\n",
      " 104 105 106 107 108 109 110 111 112 113 114 115 116 117 118 119 120 121\n",
      " 122 123 124 125 126 128 129 130 131 132 133 134 135 136 137 138 139 140\n",
      " 141]\n",
      "In Data Processing Reference Sequence (shape= (109,) ): \n",
      " ['Y' 'R' 'E' 'V' 'N' 'Y' 'L' 'H' 'E' 'N' 'S' 'I' 'E' 'K' 'L' 'S' 'P' 'H'\n",
      " 'I' 'K' 'M' 'L' 'S' 'S' 'A' 'L' 'N' 'R' 'L' 'V' 'A' 'A' 'F' 'R' 'G' 'T'\n",
      " 'V' 'Y' 'G' 'I' 'T' 'I' 'A' 'R' 'L' 'Y' 'H' 'F' 'D' 'T' 'G' 'G' 'R' 'Y'\n",
      " 'V' 'E' 'P' 'A' 'F' 'M' 'T' 'T' 'R' 'I' 'K' 'D' 'Q' 'V' 'F' 'E' 'P' 'G'\n",
      " 'I' 'A' 'F' 'Q' 'I' 'S' 'L' 'K' 'R' 'G' 'A' 'D' 'I' 'S' 'G' 'S' 'S' 'Q'\n",
      " 'A' 'P' 'S' 'E' 'E' 'I' 'M' 'L' 'P' 'M' 'M' 'S' 'E' 'F' 'V' 'I' 'E' 'H'\n",
      " 'A']\n",
      "\n",
      "\n",
      "\n",
      "Preprocessed reference Sequence:  [19 14  3 17 11 19  9  6  3 11 15  7  3  8  9 15 12  6  7  8 10  9 15 15\n",
      "  0  9 11 14  9 17  0  0  4 14  5 16 17 19  5  7 16  7  0 14  9 19  6  4\n",
      "  2 16  5  5 14 19 17  3 12  0  4 10 16 16 14  7  8  2 13 17  4  3 12  5\n",
      "  7  0  4 13  7 15  9  8 14  5  0  2  7 15  5 15 15 13  0 12 15  3  3  7\n",
      " 10  9 12 10 10 15  3  4 17  7  3  6  0]\n",
      "Original Sequence length:  912\n",
      "After removing bad sequences, tpdb is now  442\n",
      "99\n",
      "After removing bad sequences, tpdb is now  348\n",
      "s after data_processing:  (367, 151)\n",
      "Reference sequence (tpdb, s_ipdb) is sequence # 348\n",
      "shape of s \n",
      "import from /data/cresswellclayec/DCA_ER/Pfam-A.full/PF03496/msa.npy\n",
      " (367, 151)\n",
      "/data/cresswellclayec/DCA_ER/Pfam-A.full/PF03496\n",
      "(367, 151) \n",
      "\n",
      "\n",
      "\n",
      "    PDB CHAIN  PDB_START  PDB_END PFAM_ACCESSION       PFAM_NAME  \\\n",
      "0  101m     A          7      113        PF00042          Globin   \n",
      "1  102l     A         24      149        PF00959  Phage_lysozyme   \n",
      "2  102m     A          7      113        PF00042          Globin   \n",
      "3  103l     A         24      151        PF00959  Phage_lysozyme   \n",
      "4  103m     A          7      113        PF00042          Globin   \n",
      "\n",
      "  AUTH_PDBRES_START AUTH_PDBRES_START_INS_CODE AUTH_PDBRES_END  \\\n",
      "0                 6                                        112   \n",
      "1                24                                        148   \n",
      "2                 6                                        112   \n",
      "3                24                                        148   \n",
      "4                 6                                        112   \n",
      "\n",
      "  AUTH_PDBRES_END_INS_CODE UNIPROT_ACCESSION  UNP_START  UNP_END  \n",
      "0                                     P02185          7      113  \n",
      "1                                     P00720         24      148  \n",
      "2                                     P02185          7      113  \n",
      "3                                     P00720         24      148  \n",
      "4                                     P02185          7      113  \n"
     ]
    }
   ],
   "source": [
    "ipdb = 0\n",
    "printing = True\n",
    "print('seq:',int(pdb[ipdb,1]))\n",
    "\n",
    "\n",
    "\n",
    "s0,cols_removed, s_index, tpdb, orig_seq_len = data_processing(data_path, pfam_id, ipdb,\\\n",
    "                gap_seqs=0.2, gap_cols=0.2, prob_low=0.004, conserved_cols=0.9, printing=printing, out_dir=processed_data_dir)\n",
    "print('\\n\\n\\nPreprocessed reference Sequence: ', s0[tpdb])\n",
    "\n",
    "# npy2fa does not remove cols this way we are as close to original as possible\n",
    "msa_outfile, ref_outfile, s0, cols_removed, s_index, tpdb, orig_seq_len  = tools.npy2fa(pfam_id, msa_npy_file, pdb_ref_file=pdb_ref_file, ipdb=ipdb, preprocess=True,gap_seqs=.2, gap_cols=.2, prob_low=.004, conserved_cols=.9, letter_format=False)\n",
    "# data processing\n",
    "\n",
    "print(s0.shape, '\\n\\n\\n')\n",
    "\n",
    "# # Load pydca trimmed datrimmed_data_outfile = '%sprotein_data/data_processing_output/MSA_%s_Trimmed.fa' % (DCA_ER_dir, pfam_id)\n",
    "# trimmed_msa_file = Path(DCA_ER_dir, 'protein_data/data_processing_output/MSA_%s_Trimmed.fa' % pfam_id)\n",
    "# s0_pydca = tools.read_FASTA(str(trimmed_msa_file), ref_index = int(pdb[ipdb,1]))\n",
    "\n",
    "# Load csv of PDB-PFAM mapping.\n",
    "#    downloaded from \n",
    "#    ftp://ftp.ebi.ac.uk/pub/databases/msd/sifts/flatfiles/csv/pdb_pfam_mapping.csv.gz\n",
    "pdb_id = pdb_df.iloc[ipdb]['pdb_id']\n",
    "pdb_chain = pdb_df.iloc[ipdb]['chain']\n",
    "\n",
    "pdb_pfam_map_file = Path('%s/protein_data/pdb_data/pdb_pfam_mapping.csv' % DCA_dir)\n",
    "pdb_map_df = pd.read_csv(pdb_pfam_map_file, sep=',', header=1)\n",
    "print(pdb_map_df.head())\n",
    "\n",
    "pdb_id_map_df = pdb_map_df.loc[pdb_map_df['PDB']==pdb_id.lower()]\n",
    "pdb_pfam_map = pdb_id_map_df.loc[pdb_id_map_df['CHAIN']==pdb_chain]\n",
    "\n",
    "# print('wtf pdb length matching S sequence length ', orig_seq_len )\n",
    "# for pdb_index, pdb_info in pdb_pfam_map.iterrows(): # loop trhough rows of dataframe\n",
    "#     pdb_len = pdb_info['PDB_END'] - pdb_info['PDB_START']\n",
    "#     print(pdb_len)\n",
    "#     if pdb_len == orig_seq_len:\n",
    "#         break\n",
    "        \n",
    "#       "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Polypeptide range for contact map:  [114, 264]\n",
      "\n",
      "\n",
      "#-----------------------#\n",
      "Generating Contact Map\n",
      "#----------------------------#\n",
      "\n",
      "['PF03496' '445' 'Q88A91_PSESM' '108' '258' '3U0J' 'B' '108' '258']\n",
      "Downloading PDB structure '3U0J'...\n",
      "peptide build of chain produced 1 elements\n",
      "original poly_seq: \n",
      " ['E', 'R', 'H', 'Y', 'S', 'T', 'G', 'Q', 'D', 'R', 'H', 'D', 'F', 'Y', 'R', 'F', 'A', 'A', 'R', 'L', 'H', 'V', 'D', 'A', 'Q', 'C', 'F', 'G', 'L', 'S', 'I', 'D', 'D', 'L', 'M', 'D', 'K', 'F', 'S', 'D', 'K', 'H', 'F', 'R', 'A', 'E', 'H', 'P', 'E', 'Y', 'R', 'D', 'V', 'Y', 'P', 'E', 'E', 'C', 'S', 'A', 'I', 'Y', 'M', 'H', 'T', 'A', 'Q', 'D', 'Y', 'S', 'S', 'H', 'L', 'V', 'R', 'G', 'E', 'I', 'G', 'T', 'P', 'L', 'Y', 'R', 'E', 'V', 'N', 'N', 'Y', 'L', 'R', 'L', 'Q', 'H', 'E', 'N', 'S', 'G', 'R', 'E', 'A', 'E', 'I', 'D', 'N', 'H', 'D', 'E', 'K', 'L', 'S', 'P', 'H', 'I', 'K', 'M', 'L', 'S', 'S', 'A', 'L', 'N', 'R', 'L', 'M', 'D', 'V', 'A', 'A', 'F', 'R', 'G', 'T', 'V', 'Y', 'R', 'G', 'I', 'R', 'G', 'D', 'L', 'D', 'T', 'I', 'A', 'R', 'L', 'Y', 'H', 'L', 'F', 'D', 'T', 'G', 'G', 'R', 'Y', 'V', 'E', 'P', 'A', 'F', 'M', 'S', 'T', 'T', 'R', 'I', 'K', 'D', 'S', 'A', 'Q', 'V', 'F', 'E', 'P', 'G', 'T', 'P', 'N', 'N', 'I', 'A', 'F', 'Q', 'I', 'S', 'L', 'K', 'R', 'G', 'A', 'D', 'I', 'S', 'G', 'S', 'S', 'Q', 'A', 'P', 'S', 'E', 'E', 'E', 'I', 'M', 'L', 'P', 'M', 'M', 'S', 'E', 'F', 'V', 'I', 'E', 'H', 'A', 'S', 'A', 'L', 'S', 'E', 'G', 'K', 'H', 'L', 'F', 'V', 'L', 'S', 'Q', 'I'] \n",
      " length:  236\n",
      "Polypeptide Ca Coord list:  236 \n",
      " length:  236\n",
      "length of pp_ca_coords_full  236\n",
      "pdb range 113, 264\n",
      "length of pp_ca_coords_full  123\n",
      "\n",
      "\n",
      "Extracting pdb-range from full list of polypeptide coordinates\n",
      "PDB-range poly_seq: \n",
      " ['I', 'K', 'M', 'L', 'S', 'S', 'A', 'L', 'N', 'R', 'L', 'M', 'D', 'V', 'A', 'A', 'F', 'R', 'G', 'T', 'V', 'Y', 'R', 'G', 'I', 'R', 'G', 'D', 'L', 'D', 'T', 'I', 'A', 'R', 'L', 'Y', 'H', 'L', 'F', 'D', 'T', 'G', 'G', 'R', 'Y', 'V', 'E', 'P', 'A', 'F', 'M', 'S', 'T', 'T', 'R', 'I', 'K', 'D', 'S', 'A', 'Q', 'V', 'F', 'E', 'P', 'G', 'T', 'P', 'N', 'N', 'I', 'A', 'F', 'Q', 'I', 'S', 'L', 'K', 'R', 'G', 'A', 'D', 'I', 'S', 'G', 'S', 'S', 'Q', 'A', 'P', 'S', 'E', 'E', 'E', 'I', 'M', 'L', 'P', 'M', 'M', 'S', 'E', 'F', 'V', 'I', 'E', 'H', 'A', 'S', 'A', 'L', 'S', 'E', 'G', 'K', 'H', 'L', 'F', 'V', 'L', 'S', 'Q', 'I'] \n",
      " length:  123\n",
      "PDB-range Polypeptide Ca Coord list:  123 \n",
      " length:  123\n",
      "\n",
      "\n",
      "\n"
     ]
    },
    {
     "ename": "IndexError",
     "evalue": "index 142 is out of bounds for axis 0 with size 123",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mIndexError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[0;32m/tmp/ipykernel_27743/2149062229.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     14\u001b[0m \u001b[0;31m# returns contact map with the appropriate columns removed..\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     15\u001b[0m \u001b[0;31m# For list of retained columns us s_index\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 16\u001b[0;31m \u001b[0mct\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mct_full\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mn_amino_full\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mpoly_seq_curated\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mtools\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcontact_map\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mpdb\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mipdb\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mpp_range\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcols_removed\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0ms_index\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mpdb_out_dir\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mpdb_out\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mprinting\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mprinting\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     17\u001b[0m \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mct\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mshape\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     18\u001b[0m \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mct_full\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mshape\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/gpfs/gsfs8/users/cresswellclayec/DCA_ER/ecc_tools.py\u001b[0m in \u001b[0;36mcontact_map\u001b[0;34m(pdb, ipdb, pp_range, cols_removed, s_index, ref_seq, printing, pdb_out_dir)\u001b[0m\n\u001b[1;32m    494\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    495\u001b[0m         \u001b[0;31m# Get curated polypeptide sequence (using cols_removed) & Get coords of curated pp seq\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 496\u001b[0;31m     \u001b[0mpoly_seq_curated\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdelete\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mpoly_seq_range\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcols_removed\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    497\u001b[0m     \u001b[0mpp_ca_coords_curated\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdelete\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mpp_ca_coords_full_range\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcols_removed\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    498\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m<__array_function__ internals>\u001b[0m in \u001b[0;36mdelete\u001b[0;34m(*args, **kwargs)\u001b[0m\n",
      "\u001b[0;32m/data/cresswellclayec/conda/envs/DCA_ER/lib/python3.9/site-packages/numpy/lib/function_base.py\u001b[0m in \u001b[0;36mdelete\u001b[0;34m(arr, obj, axis)\u001b[0m\n\u001b[1;32m   4478\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   4479\u001b[0m             \u001b[0mkeep\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mones\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mN\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdtype\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mbool\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 4480\u001b[0;31m             \u001b[0mkeep\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mobj\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mFalse\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   4481\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   4482\u001b[0m         \u001b[0mslobj\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0maxis\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mkeep\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mIndexError\u001b[0m: index 142 is out of bounds for axis 0 with size 123"
     ]
    }
   ],
   "source": [
    "## Generate and Plot Contact Map from PDB coordinates!\n",
    "# Check that pdb--pfam mapping is unique\n",
    "if pdb_pfam_map.shape[0] > 1:\n",
    "    print('Unable to get unique PDB-->Pfam mapping')\n",
    "    print(pdb_pfam_map)\n",
    "\n",
    "# pp_range = [pdb_info['PDB_START'], pdb_info['PDB_END']]\n",
    "pp_range = [pdb_pfam_map.iloc[0]['PDB_START'], pdb_pfam_map.iloc[0]['PDB_END']]\n",
    "print('Polypeptide range for contact map: ', pp_range)\n",
    "\n",
    "\n",
    "pdb_out = \"%s/protein_data/pdb_data\" % DCA_dir\n",
    "# Directory for storing PDB data locally\n",
    "# returns contact map with the appropriate columns removed..\n",
    "# For list of retained columns us s_index\n",
    "ct, ct_full, n_amino_full, poly_seq_curated = tools.contact_map(pdb, ipdb, pp_range, cols_removed, s_index, pdb_out_dir=pdb_out, printing=printing)\n",
    "print(ct.shape)\n",
    "print(ct_full.shape)\n",
    "\n",
    "\n",
    "ct_mat = ct\n",
    "# tools.hide_toggle()\n",
    "print('Contact matrix shape:', ct_mat.shape)\n",
    "\n",
    "plt.tight_layout(h_pad=1, w_pad=1.5)\n",
    "plt.title('Contact Matrix')\n",
    "plt.imshow(ct_mat,cmap='rainbow',origin='lower')\n",
    "plt.xlabel('i')\n",
    "plt.ylabel('j')\n",
    "plt.clim(0,25)\n",
    "plt.colorbar(fraction=0.045, pad=0.05)\n",
    "# 'I'' 'Q' 'S' 'R' 'V' 'E' 'T' 'V' 'P' 'Y' 'L' 'F' 'C' 'R' 'S' 'E' 'E' 'V'\n",
    "#  'R' 'P' 'A' 'G' 'M' 'V' 'W' 'Y' 'S' 'I' 'L' 'K' 'D' 'T' 'K' 'I' 'T' 'C'\n",
    "#  'E' 'E' 'K' 'M' 'V' 'S' 'M' 'A' 'R' 'N' 'T\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Plot the contact Map of the imported Protein structure"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Now we want to use Expectation Reflection to infer the \"calculate the weights \" \n",
    "  * Infer interactions between residues (R-part of ammino acid) in proteins from Multiple Sequence Alignment (MSA) data\n",
    "  * More understanding required on meaning/interpretation of weights\n",
    "  \n",
    "1. We can start by defining some important values of our protein structures\n",
    "    * It is important to define 'i1i2' vector to keep track of OneHot index positions (i) coresonding to the beginning and end of a given amino acid position (this acts as a key between the OneHot matrix (s) and the original sequence matrix (s0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of residue positions: 151\n",
      "Number of sequences: 367\n",
      "Number of different amino acids at each position [ 2  2  2 11 19 19 10  2 18 19  8  1  1  1 18 17 19 19  2  2  2  2  2 19\n",
      "  2  2  2  1 19 19 17 20 18 19 14 18 20  8 12 14 12  7 18 18 20  2  2 16\n",
      " 16 17 17 16 17 19 12 18  2 16 16  2  2  2  2  2 19 20 19 19 20 20 19  2\n",
      " 19 20 15 11 18 18 16 18 16 17 13  8 11  2  8  6 15 19 18 18  2  2 19 20\n",
      " 18 20 20 19  2  2  2  2 20 21 16 20  9 20 15 19 17 15 17 19 13 17 18 17\n",
      " 16 19 20 19 20  4 17  2  9  4  8 15 17 13 16 19  9 18  9 14 12 13  2  2\n",
      "  2  2  2  2  2  2  2]\n",
      "Total number of variables 1829\n"
     ]
    }
   ],
   "source": [
    "# number of positions\n",
    "n_var = s0.shape[1]\n",
    "n_seq = s0.shape[0]\n",
    "\n",
    "print(\"Number of residue positions:\",n_var)\n",
    "print(\"Number of sequences:\",n_seq)\n",
    "\n",
    "# number of aminoacids at each position\n",
    "mx = np.array([len(np.unique(s0[:,i])) for i in range(n_var)])\n",
    "#mx = np.array([m for i in range(n_var)])\n",
    "print(\"Number of different amino acids at each position\",mx)\n",
    "\n",
    "mx_cumsum = np.insert(mx.cumsum(),0,0)\n",
    "i1i2 = np.stack([mx_cumsum[:-1],mx_cumsum[1:]]).T \n",
    "# print(\"(Sanity Check) Column indices of first and (\",i1i2[0],\") and last (\",i1i2[-1],\") positions\")\n",
    "# print(\"(Sanity Check) Column indices of second and (\",i1i2[1],\") and second to last (\",i1i2[-2],\") positions\")\n",
    "\n",
    "\n",
    "# number of variables\n",
    "mx_sum = mx.sum()\n",
    "print(\"Total number of variables\",mx_sum)\n",
    "\n",
    "# number of bias term\n",
    "n_linear = mx_sum - n_var"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "2.  The above indices help to divide all the possible amino acids at different positions to binary choices which will be implemented wit 1 of K notation (OneHot Encoder)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "onehot_encoder = OneHotEncoder(sparse=False,categories='auto')\n",
    "# s is OneHot encoder format, s0 is original sequnce matrix\n",
    "s = onehot_encoder.fit_transform(s0)\n",
    "# print(\"Amino Acid sequence Matrix\\n\",s0)\n",
    "# print(\"OneHot sequence Matrix\\n\",s)\n",
    "# print(\"An individual element of the OneHot sequence Matrix (size:\",\n",
    "#      s.shape,\") --> \",s[0], \" has length \",s[0].shape)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### OneHot Format example\n",
    "* Given 3 sequence poitions each with 3 different amino acids:\n",
    "    * Notation: The 3<sup>rd</sup> amino acid type at the 2<sup>nd</sup> sequence position is denoted as R23<br/>\n",
    "* OneHot converts the sequence [R11 R22 R33] &rightarrow; [1 0 0 0 1 0 0 0 1]  \n",
    "  \n",
    "  \n",
    "3. We next want to define our local field parameters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define wight matrix with variable for each possible amino acid at each sequence position\n",
    "w = np.zeros((mx.sum(),mx.sum())) \n",
    "h0 = np.zeros(mx.sum())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Apply Expectation Reflection\n",
    "4. We can now apply expectation reflection to infer the connection between postion residues"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "import numpy as np\n",
    "from scipy import linalg\n",
    "from sklearn.preprocessing import OneHotEncoder\n",
    "import expectation_reflection as ER\n",
    "from direct_info import direct_info\n",
    "from joblib import Parallel, delayed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "pycharm": {
     "is_executing": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "run time: 117.73817592859268\n",
      "run time: 67.10801172628999\n"
     ]
    }
   ],
   "source": [
    "# Expectation Reflection\n",
    "#=========================================================================================\n",
    "def predict_w(s,i0,i1i2,niter_max,l2):\n",
    "    #print('i0:',i0)\n",
    "    i1,i2 = i1i2[i0,0],i1i2[i0,1]\n",
    "\n",
    "    x = np.hstack([s[:,:i1],s[:,i2:]])\n",
    "    y = s[:,i1:i2]\n",
    "\n",
    "    h01,w1 = ER.fit(x,y,niter_max,l2)\n",
    "\n",
    "    return h01,w1\n",
    "\n",
    "\n",
    "if os.path.exists(\"%s_w.npy\" % pfam_id) and not create_new:\n",
    "# if 0:\n",
    "    w = np.load(\"%s_w.npy\" % pfam_id)\n",
    "else:\n",
    "    #-------------------------------\n",
    "    # parallel\n",
    "    start_time = timeit.default_timer()\n",
    "    #res = Parallel(n_jobs = 4)(delayed(predict_w)\\\n",
    "    res = Parallel(n_jobs = n_jobs)(delayed(predict_w)\\\n",
    "    #res = Parallel(n_jobs = 32)(delayed(predict_w)\\\n",
    "            (s,i0,i1i2,niter_max=10,l2=100.0)\\\n",
    "            for i0 in range(n_var))\n",
    "\n",
    "    run_time = timeit.default_timer() - start_time\n",
    "    print('run time:',run_time)\n",
    "    ## This above line seems wrong, seems like the following for loop should be moved up?? not sure if this is some \n",
    "    ## python implementation or just wrong\n",
    "    #----------------niter_max,l2)\n",
    "\n",
    "    #-------------------------------\n",
    "    # parallel\n",
    "    start_time = timeit.default_timer()\n",
    "    #res = Parallel(n_jobs = 4)(delayed(predict_w)\\\n",
    "    #res = Parallel(n_jobs = 8)(delayed(predict_w)\\\n",
    "    res = Parallel(n_jobs = 32)(delayed(predict_w)\\\n",
    "            (s,i0,i1i2,niter_max=10,l2=100.0)\\\n",
    "            for i0 in range(n_var))\n",
    "\n",
    "    run_time = timeit.default_timer() - start_time\n",
    "    print('run time:',run_time)\n",
    "    ## This above line seems wrong, seems like the following for loop should be moved up?? not sure if this is some \n",
    "    ## python implementation or just wrong\n",
    "    #-------------------------------\n",
    "    for i0 in range(n_var):\n",
    "        i1,i2 = i1i2[i0,0],i1i2[i0,1]\n",
    "\n",
    "        h01 = res[i0][0]\n",
    "        w1 = res[i0][1]\n",
    "\n",
    "        h0[i1:i2] = h01\n",
    "        w[:i1,i1:i2] = w1[:i1,:]\n",
    "        w[i2:,i1:i2] = w1[i1:,:]\n",
    "\n",
    "    # make w symmetric\n",
    "    w = (w + w.T)/2.\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "5. Now we want to plot a heat map of our sequence weights"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "pycharm": {
     "is_executing": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Dimensions of w:  (1829, 1829)\n"
     ]
    }
   ],
   "source": [
    "# Verify that w is symmetric (sanity test)\n",
    "print(\"Dimensions of w: \",w.shape)\n",
    "w_file = \"%s_w.npy\" % pfam_id\n",
    "np.save(w_file, w)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "6. Plot relationship between sequence positions via Direct information"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "pycharm": {
     "is_executing": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[[0.         0.00028071 0.00028072 ... 0.00028081 0.00028073 0.00028072]\n",
      " [0.00028071 0.         0.0002807  ... 0.00028079 0.00028071 0.0002807 ]\n",
      " [0.00028072 0.0002807  0.         ... 0.0002808  0.00028071 0.00028071]\n",
      " ...\n",
      " [0.00028081 0.00028079 0.0002808  ... 0.         0.0002808  0.0002808 ]\n",
      " [0.00028073 0.00028071 0.00028071 ... 0.0002808  0.         0.00028071]\n",
      " [0.00028072 0.0002807  0.00028071 ... 0.0002808  0.00028071 0.        ]]\n",
      "(151, 151)\n",
      "151\n",
      "[  0   1   2   3   4   5   6   7   8   9  10  11  12  13  14  15  16  17\n",
      "  18  19  20  21  22  23  24  25  26  27  28  29  30  31  32  33  34  35\n",
      "  36  37  38  39  40  41  42  43  44  45  46  47  48  49  50  51  52  53\n",
      "  54  55  56  57  58  59  60  61  62  63  64  65  66  67  68  69  70  71\n",
      "  72  73  74  75  76  77  78  79  80  81  82  83  84  85  86  87  88  89\n",
      "  90  91  92  93  94  95  96  97  98  99 100 101 102 103 104 105 106 107\n",
      " 108 109 110 111 112 113 114 115 116 117 118 119 120 121 122 123 124 125\n",
      " 126 127 128 129 130 131 132 133 134 135 136 137 138 139 140 141 142 143\n",
      " 144 145 146 147 148 149 150]\n",
      "scores for 11325 pairs \n",
      "DI shape (full size) (151, 151)\n",
      "DI shape (scores removed) (109, 109) \n",
      "(should be same as ER di shape..)\n"
     ]
    }
   ],
   "source": [
    "di = direct_info(s0,w)\n",
    "print(di)\n",
    "print(di.shape)\n",
    "print(len(s_index))\n",
    "\n",
    "from ecc_tools import score_APC\n",
    "sequence_length = pp_range[1] - pp_range[0]\n",
    "\n",
    "# RunE ER di scores through Average Product Correction (to combat influences of phylogeny)\n",
    "er_APC = score_APC(di, sequence_length, s_index)\n",
    "di_apc = tools.di_dict2mat(er_APC, s_index, cols_removed = cols_removed)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Plotting ER and DCA\n",
    "### Plot ER"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "pycharm": {
     "is_executing": true
    }
   },
   "outputs": [],
   "source": [
    "#print(\"Direct Information from Expectation reflection:\\n\",di)\n",
    "print('ER DI shape: ' , di.shape)\n",
    "ER_di = np.delete(di, cols_removed,0)\n",
    "ER_di = np.delete(ER_di, cols_removed,1)\n",
    "\n",
    "print('Final ER DI shape (cols removed): ', ER_di.shape)\n",
    "\n",
    "ER_di_nodiag = no_diag(ER_di, 5, s_index)\n",
    "#print(ER_di_nodiag)\n",
    "#ER_di = ER_di_nodiag"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "pycharm": {
     "is_executing": true
    }
   },
   "outputs": [],
   "source": [
    "plt.title('direct information')\n",
    "plt.imshow(ER_di,cmap='rainbow',origin='lower')\n",
    "plt.xlabel('i')\n",
    "plt.ylabel('j')\n",
    "plt.clim(0,0.01)\n",
    "plt.colorbar(fraction=0.045, pad=0.05)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "pycharm": {
     "is_executing": true
    }
   },
   "outputs": [],
   "source": [
    "from ecc_tools import roc_curve\n",
    "# find optimal threshold of distance\n",
    "ct_thres = np.linspace(4.,6.,18,endpoint=True)\n",
    "n = ct_thres.shape[0]\n",
    "\n",
    "\n",
    "print(ct.shape)\n",
    "# Initialize plotting\n",
    "iplot = [1,3,5,7,9,11,13]\n",
    "plt.figure(figsize=(9.0,3.2))\n",
    "\n",
    "# Initalize ROC-curve tile\n",
    "plt.subplot2grid((1,4),(0,0))\n",
    "plt.title('ROC various thres')\n",
    "plt.plot([0,1],[0,1],'k--')\n",
    "\n",
    "# Find ct_threshold (2-6 Angstrom) which gives best contact prediction for ER\n",
    "auc_ER = np.zeros(n)\n",
    "for i in range(n):\n",
    "    try:\n",
    "        p,tp,fp = roc_curve(ct_mat, ER_di, ct_thres[i])\n",
    "        if i in iplot:\n",
    "            plt.plot(fp,tp,label='thres = %3.2f'%ct_thres[i])\n",
    "        auc_ER[i] = tp.sum()/tp.shape[0]\n",
    "    except(RuntimeWarning):\n",
    "        auc_ER[i] = 0\n",
    "\n",
    "# Get ER method's best contact prediction\n",
    "i0_ER = np.argmax(auc_ER)\n",
    "print('ER auc max:',ct_thres[i0_ER],auc_ER[i0_ER])\n",
    "p0_ER,tp0_ER,fp0_ER = tools.roc_curve(ct_mat,ER_di,ct_thres[i0_ER])\n",
    "\n",
    "# ROC-curve tile settings\n",
    "plt.xlim([0,1])\n",
    "plt.ylim([0,1])\n",
    "plt.xlabel('False Positive Rate')\n",
    "plt.ylabel('True Positive Rate')\n",
    "plt.legend()\n",
    "\n",
    "\n",
    "# Plot ROC, AUC and Precision for best ER contact prediction\n",
    "# AUC\n",
    "plt.subplot2grid((1,4),(0,1))\n",
    "plt.title('AUC max = %f' %(auc_ER[i0_ER]))\n",
    "plt.plot([ct_thres.min(),ct_thres.max()],[0.5,0.5],'k--')\n",
    "plt.plot(ct_thres,auc_ER,'k-')\n",
    "plt.plot(ct_thres[i0_ER],auc_ER[i0_ER], 'b*', markersize=10.)\n",
    "plt.xlim([ct_thres.min(),ct_thres.max()])\n",
    "plt.ylim([0,auc_ER.max()+0.1])\n",
    "plt.xlabel('distance threshold')\n",
    "plt.ylabel('AUC')\n",
    "# ROC\n",
    "plt.subplot2grid((1,4),(0,2))\n",
    "plt.title('ROC at thres = %3.2f'%(ct_thres[i0_ER]))\n",
    "plt.plot(fp0_ER,tp0_ER,'b-')\n",
    "plt.plot([0,1],[0,1],'k--')\n",
    "plt.xlim([0,1])\n",
    "plt.ylim([0,1])\n",
    "plt.xlabel('False Positive Rate')\n",
    "plt.ylabel('True Positive Rate')\n",
    "# Precision-Recall\n",
    "plt.subplot2grid((1,4),(0,3))\n",
    "plt.title('Precision-Recall Curve')\n",
    "print('\\nContact threshold: ', ct_thres[i0_ER])\n",
    "plt.plot( p0_ER,tp0_ER / (tp0_ER + fp0_ER), 'b-', label='thres = %s'%ct_thres[i0_ER])\n",
    "plt.xlabel('Recall (Sensitivity - P)')\n",
    "plt.ylabel('Precision (PPV)')\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Plot DCA "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "pycharm": {
     "is_executing": true
    }
   },
   "outputs": [],
   "source": [
    "from inference_dca import direct_info_dca\n",
    "print(s0.shape)\n",
    "seq_wt_file = None\n",
    "seq_wt_file = '%s/protein_data/data_processing_output/seq_weight_%s.npy' % (DCA_ER_dir, pfam_id)\n",
    "\n",
    "# ----------- DCA DI (MF) calculation --------------------------------------------- #\n",
    "\n",
    "mf_di, fi, fij, c, cinv, w, w2d, fi_pydca, fij_pydca, c_pydca, c_inv_pydca, \\\n",
    "w_pydca, w2d_pydcak, di_pydca, ma_inv,seq_ints\\\n",
    "= direct_info_dca(s0, seq_wt_outfile=seq_wt_file)\n",
    "print('c[0]:',c[0])\n",
    "print('c_pydca[0]', c_pydca[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "pycharm": {
     "is_executing": true
    }
   },
   "outputs": [],
   "source": [
    "MF_di = np.delete(mf_di, cols_removed,0)\n",
    "MF_di = np.delete(MF_di, cols_removed,1)\n",
    "MF_hybrid_di = np.delete(di_pydca, cols_removed,0)\n",
    "MF_hybrid_di = np.delete(MF_hybrid_di, cols_removed,1)\n",
    "# --------------------------------------------------------------------------------- #\n",
    "print(MF_di.shape)\n",
    "plt.title('Tai MF DI')\n",
    "plt.imshow(MF_di,cmap='rainbow',origin='lower')\n",
    "plt.xlabel('i')\n",
    "plt.ylabel('j')\n",
    "plt.clim(0.,.01)\n",
    "plt.colorbar(fraction=0.045, pad=0.05)\n",
    "\n",
    "plt.figure()\n",
    "print(MF_hybrid_di.shape)\n",
    "plt.title('Hybrid MF DI')\n",
    "plt.imshow(MF_hybrid_di,cmap='rainbow',origin='lower')\n",
    "plt.xlabel('i')\n",
    "plt.ylabel('j')\n",
    "plt.clim(2.,0.)\n",
    "plt.colorbar(fraction=0.045, pad=0.05)\n",
    "print('cinv[0]', cinv[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "pycharm": {
     "is_executing": true
    }
   },
   "outputs": [],
   "source": [
    "# Run our version of MF di scores through Average Product Correction (to combat influences of phylogeny)\n",
    "# mf_APC = score_APC(MF_di, sequence_length, s_index)\n",
    "# MF_di = tools.di_dict2mat(mf_APC, cols_removed, s_index)\n",
    "\n",
    "# Initialize plotting\n",
    "iplot = [1,3,5,7,9,11,13]\n",
    "plt.figure(figsize=(9.0,3.2))\n",
    "\n",
    "plt.subplot2grid((1,4),(0,0))\n",
    "plt.title('ROC various thres')\n",
    "plt.plot([0,1],[0,1],'k--')\n",
    "\n",
    "# find optimal threshold of distance\n",
    "auc_MF = np.zeros(n)\n",
    "\n",
    "for i in range(n):\n",
    "    try:\n",
    "        p,tp,fp = roc_curve(ct_mat, MF_di, ct_thres[i])\n",
    "        if i in iplot:\n",
    "            plt.plot(fp,tp,label='thres = %3.2f'%ct_thres[i])\n",
    "        auc_MF[i] = tp.sum()/tp.shape[0]\n",
    "    except(RuntimeWarning):\n",
    "        auc_MF[i] = 0\n",
    "plt.xlim([0,1])\n",
    "plt.ylim([0,1])\n",
    "plt.xlabel('False Positive Rate')\n",
    "plt.ylabel('True Positive Rate')\n",
    "        \n",
    "i0_MF = np.argmax(auc_MF)\n",
    "print('auc max:',ct_thres[i0_MF],auc_MF[i0_MF])\n",
    "p0_MF,tp0_MF,fp0_MF = tools.roc_curve(ct,MF_di,ct_thres[i0_MF])\n",
    "\n",
    "\n",
    "plt.subplot2grid((1,4),(0,1))\n",
    "plt.title('AUC max = %f' %(auc_MF[i0_MF]))\n",
    "plt.plot([ct_thres.min(),ct_thres.max()],[0.5,0.5],'k--')\n",
    "plt.plot(ct_thres,auc_MF,'r-')\n",
    "plt.plot(ct_thres[i0_MF],auc_MF[i0_MF], 'r*', markersize=10.)\n",
    "plt.xlim([ct_thres.min(),ct_thres.max()])\n",
    "plt.ylim([0,auc_MF.max()+0.1])\n",
    "plt.xlabel('distance threshold')\n",
    "plt.ylabel('AUC')\n",
    "\n",
    "plt.subplot2grid((1,4),(0,2))\n",
    "plt.title('ROC at thres = %3.2f'%(ct_thres[i0_MF]))\n",
    "plt.plot(fp0_MF,tp0_MF,'r-')\n",
    "plt.plot([0,1],[0,1],'k--')\n",
    "plt.xlim([0,1])\n",
    "plt.ylim([0,1])\n",
    "plt.xlabel('False Positive Rate')\n",
    "plt.ylabel('True Positive Rate')\n",
    "\n",
    "plt.subplot2grid((1,4),(0,3))\n",
    "plt.title('Precision-Recall Curve')\n",
    "plt.plot(p0_MF, tp0_MF / (tp0_MF + fp0_MF), 'r-', label='thres = %s'%ct_thres[i0_MF])\n",
    "plt.xlabel('Recall (Sensitivity - P)')\n",
    "plt.ylabel('Precision (PPV)')\n",
    "\n",
    "plt.tight_layout(h_pad=1, w_pad=1.5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Load PYDCA Results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def slice_couplings(couplings = None, site_pair = None, num_site_states=None):\n",
    "    \"\"\"Returns couplings corresponding to site pair (i, j). Note that the\n",
    "    the couplings involving gaps are included, but they are set to zero.\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "        couplings : np.array\n",
    "            A 2d numpy array of couplings. It has a shape of (L(q-1), L(q-1))\n",
    "            where L and q are the length of sequences in alignment data and total\n",
    "            number of standard residues plus gap.\n",
    "        site_pair : tuple\n",
    "            A tuple of site pairs. Example (0, 1), (0, L-1), ..., (L-2, L-1).\n",
    "        num_site_states : int\n",
    "            The value of q.\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "        couplings_ij : np.array\n",
    "            A2d numpy array of shape (q, q) containing the couplings. Note that\n",
    "            couplings_ij[q, :] and couplings[:, q] are set to zero.\n",
    "    \"\"\"\n",
    "    q = num_site_states\n",
    "    couplings_ij = np.zeros((q, q), dtype = np.float64)\n",
    "    row_begin = site_pair[0] * (q - 1)\n",
    "    row_end = row_begin + q - 1\n",
    "    column_begin = site_pair[1] * (q -1)\n",
    "    column_end = column_begin + q - 1\n",
    "    couplings_ij[:q-1, :q-1] = couplings[row_begin:row_end, column_begin:column_end]\n",
    "    return couplings_ij\n",
    "\n",
    "# -------------------- Load PYDCA Data ------------------------------------------------------------------- #\n",
    "DCA_ER_dir = '/home/evan/temp_DCA_ER_pydcaMF/DCA_ER/'\n",
    "DCA_ER_dir = '/home/evan/PycharmProjects/DCA_ER/'\n",
    "DCA_ER_dir = '/home/ecresswell/DCA_ER/'\n",
    "DCA_ER_dir = '/data/cresswellclayec/DCA_ER/'\n",
    "\n",
    "out_dir = '%sprotein_data/di/' % DCA_ER_dir\n",
    "method = \"DI\"\n",
    "\n",
    "\n",
    "preprocess_pydca = False\n",
    "if preprocess_pydca:\n",
    "    # Load pydca score pseudo-dictionaries and generate di matrices from them..\n",
    "    mf_score_file = '%s%s_%s_pydca_mf_preproc.npy' % (out_dir, pfam_id, method)\n",
    "    plm_score_file = '%s%s_%s_pydca_plm_preproc.npy' % (out_dir, pfam_id, method)\n",
    "else:\n",
    "    # Load pydca score pseudo-dictionaries and generate di matrices from them..\n",
    "    mf_score_file = '%s%s_%s_pydca_mf.npy' % (out_dir, pfam_id, method)\n",
    "    plm_score_file = '%s%s_%s_pydca_plm.npy' % (out_dir, pfam_id, method)  \n",
    "# list of di scores length of: seq_len choose 2 \n",
    "# (ie for PF00186 with sequence length 160, this should be length 160 choose 2 = 12720)\n",
    "mf_scores = np.load(mf_score_file, allow_pickle=True)\n",
    "plm_scores = np.load(plm_score_file, allow_pickle=True)\n",
    "\n",
    "# Load pydca data for comparison\n",
    "reg_fij = np.load('%s%s_pydca_fij_preproc.npy' % (out_dir, pfam_id))\n",
    "reg_fi = np.load('%s%s_pydca_fi_preproc.npy' % (out_dir, pfam_id))\n",
    "corr_mat = np.load('%s%s_pydca_corr_preproc.npy' % (out_dir, pfam_id))                                                                             \n",
    "couplings = np.load('%s%s_pydca_couplings_preproc.npy' % (out_dir, pfam_id))        \n",
    "fields_ij = np.load('%s%s_pydca_fields_preproc.npy' % (out_dir, pfam_id))\n",
    "\n",
    "# print(corr_mat[1][:20])\n",
    "# print(c[1][:20])\n",
    "# print(corr_mat.shape)\n",
    "# print(c.shape)\n",
    "# -------------------------------------------------------------------------------------------------------- #\n",
    "# plot difference between coupling matrices (PYDCA MF and our MF)\n",
    "\n",
    "from inference_dca import interactions, direct_info, mapkey\n",
    "l = s0.shape[1] # l --> number of aa in each sequence\n",
    "print('aa sequence length l = %d' % l)\n",
    "\n",
    "q = 21\n",
    "\n",
    "\n",
    "print(couplings.shape) # 160 pos x 20 aa\n",
    "print(w2d.shape) # 137 pos x 21 aa\n",
    "print(cinv.shape)\n",
    "print(c.shape)\n",
    "print(w[0][5][0])\n",
    "print(w.shape)\n",
    "\n",
    "w_2D = np.zeros(couplings.shape)\n",
    "w_vals = []\n",
    "for i in range(l):\n",
    "    for j in range(i, l):\n",
    "        for alpha in range(q-1):\n",
    "            row = mapkey(i,alpha,q)\n",
    "            for beta in range(q-1):\n",
    "                col=mapkey(j,beta,q)\n",
    "                try:\n",
    "                    w_2D[row, col] = w[i][j][alpha][beta]\n",
    "                    w_2D[col, row] = w[i][j][alpha][beta]\n",
    "                    w_vals.append(couplings[row,col])\n",
    "                except(IndexError):\n",
    "                    print(i, j, alpha, beta)\n",
    "                    print(w_2D)\n",
    "print('min value in w: ', min(w_vals))\n",
    "print('max value in w: ', max(w_vals))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Compare Optimal MF and ER\n",
    "* Taking the optimal threshold for each gives comparable accuracy\n",
    "* ER Has higher accuracy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "# print(reg_fi.shape)\n",
    "# print(np.delete(reg_fi, cols_removed, axis=0))\n",
    "# print(np.delete(reg_fi, cols_removed, axis=0).shape)\n",
    "# print(fi)\n",
    "# print(fi.shape)\n",
    "\n",
    "# print(len(mf_fn_apc))\n",
    "# print(MF_di.shape)\n",
    "# print(mf_fn_apc[:5])\n",
    "pydca_MF_di = tools.di_dict2mat(mf_scores, s_index, cols_removed=cols_removed, aa_index_correction=True)\n",
    "pydca_PLM_di = tools.di_dict2mat(plm_scores, s_index, cols_removed=cols_removed, aa_index_correction=True)\n",
    "\n",
    "# PYDCA_MF_di = np.delete(pydca_MF_di, cols_removed,0)\n",
    "# PYDCA_MF_di = np.delete(PYDCA_MF_di, cols_removed,1)\n",
    "\n",
    "# PYDCA_PLM_di = np.delete(pydca_PLM_di, cols_removed,0)\n",
    "# PYDCA_PLM_di = np.delete(PYDCA_PLM_di, cols_removed,1)\n",
    "\n",
    "print('\\n',pydca_PLM_di.shape)\n",
    "print(pydca_MF_di.shape)\n",
    "\n",
    "print(max(s_index))\n",
    "\n",
    "plt.title('PYDCA MF direct information')\n",
    "plt.imshow(pydca_MF_di,cmap='rainbow',origin='lower')\n",
    "plt.xlabel('i')\n",
    "plt.ylabel('j')\n",
    "plt.clim(0.,.01)\n",
    "plt.colorbar(fraction=0.045, pad=0.05)\n",
    "plt.figure()\n",
    "plt.title('PYDCA PLM direct information')\n",
    "plt.imshow(pydca_PLM_di,cmap='rainbow',origin='lower')\n",
    "plt.xlabel('i')\n",
    "plt.ylabel('j')\n",
    "plt.clim(0.,.01)\n",
    "plt.colorbar(fraction=0.045, pad=0.05)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "non_triv_cinv = np.nonzero(cinv)\n",
    "print(non_triv_cinv)\n",
    "non_triv_c = np.nonzero(c)\n",
    "print(non_triv_c)\n",
    "eig_vals, eig_vecs = np.linalg.eig(c)\n",
    "hybrid_eig_vals, hybrid_eig_vecs = np.linalg.eig(c_pydca)\n",
    "pydca_eig_vals, pydca_eig_vegs = np.linalg.eig(corr_mat)\n",
    "eig_diff = [abs(eig_vals[i] - pydca_eig_vals[i]) for i in range(len(eig_vals))]\n",
    "print('max, mean, std dev eig diff: ', max(eig_diff), np.mean(eig_diff), np.std(eig_diff))\n",
    "count = 0\n",
    "for diff in eig_diff:\n",
    "    if diff > .001:\n",
    "        count = count + 1\n",
    "print('%d eigenvalues with diff greater than .001' % count)\n",
    "print('Tais min eig: ', min(eig_vals))\n",
    "print('Hybrid min eig: ', min(hybrid_eig_vals))\n",
    "print('PYDCAs min eig: ', min(pydca_eig_vals))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print (len(eig_diff))\n",
    "plt.figure(figsize=(15.0,12))\n",
    "\n",
    "# Compare Correlations\n",
    "plt.subplot2grid((3,3),(0,0))\n",
    "\n",
    "plt.plot([0,1],[0,1],'k--')\n",
    "plt.title('Tai correlation\\n(max: %f, min: %f)' % (np.amax(c), np.amin(c)))\n",
    "plt.imshow(c,cmap='rainbow',origin='lower')\n",
    "plt.xlabel('i')\n",
    "plt.ylabel('j')\n",
    "plt.clim(-.0007,.001)\n",
    "plt.colorbar(fraction=0.045, pad=0.05)\n",
    "\n",
    "plt.subplot2grid((3,3),(0,1))\n",
    "plt.title('PYDCA correlation\\n(max: %f, min: %f)' % (np.amax(corr_mat), np.amin(corr_mat)))\n",
    "plt.imshow(corr_mat,cmap='rainbow',origin='lower')\n",
    "plt.xlabel('i')\n",
    "plt.ylabel('j')\n",
    "plt.clim(-.0007,.001)\n",
    "plt.colorbar(fraction=0.045, pad=0.05)   \n",
    "\n",
    "plt.subplot2grid((3,3),(0,2))\n",
    "plt.title('Correlation Diff')\n",
    "plt.imshow(abs(corr_mat - c),cmap='rainbow',origin='lower')\n",
    "plt.xlabel('i')\n",
    "plt.ylabel('j')\n",
    "plt.clim(0,.001)\n",
    "plt.colorbar(fraction=0.045, pad=0.05)\n",
    "\n",
    "\n",
    "# Compare Weights\n",
    "plt.subplot2grid((3,3),(1,0))\n",
    "\n",
    "plt.plot([0,1],[0,1],'k--')\n",
    "plt.title('Tai weights')\n",
    "plt.imshow(w_2D,cmap='rainbow',origin='lower')\n",
    "plt.xlabel('i')\n",
    "plt.ylabel('j')\n",
    "plt.clim(-.1,1.)\n",
    "plt.colorbar(fraction=0.045, pad=0.05)\n",
    "\n",
    "plt.subplot2grid((3,3),(1,1))\n",
    "plt.title('PYDCA weights')\n",
    "plt.imshow(couplings,cmap='rainbow',origin='lower')\n",
    "plt.xlabel('i')\n",
    "plt.ylabel('j')\n",
    "plt.clim(-.1,1.)\n",
    "plt.colorbar(fraction=0.045, pad=0.05)   \n",
    "\n",
    "plt.subplot2grid((3,3),(1,2))\n",
    "plt.title('Diff')\n",
    "plt.imshow(abs(couplings - w_2D),cmap='rainbow',origin='lower')\n",
    "plt.xlabel('i')\n",
    "plt.ylabel('j')\n",
    "plt.clim(0,.2)\n",
    "plt.colorbar(fraction=0.045, pad=0.05)\n",
    "\n",
    "\n",
    "# Compare DIs \n",
    "plt.subplot2grid((3,3),(2,0))\n",
    "\n",
    "plt.plot([0,1],[0,1],'k--')\n",
    "plt.title('Tai MF DI')\n",
    "plt.imshow(MF_di,cmap='rainbow',origin='lower')\n",
    "plt.xlabel('i')\n",
    "plt.ylabel('j')\n",
    "plt.clim(0.,.01)\n",
    "plt.colorbar(fraction=0.045, pad=0.05)\n",
    "\n",
    "plt.subplot2grid((3,3),(2,1))\n",
    "plt.title('PYDCA MF DI')\n",
    "plt.imshow(pydca_MF_di,cmap='rainbow',origin='lower')\n",
    "plt.xlabel('i')\n",
    "plt.ylabel('j')\n",
    "plt.clim(0,.01)\n",
    "plt.colorbar(fraction=0.045, pad=0.05)   \n",
    "plt.savefig('taiVSpydca_colorplot.pdf')\n",
    "\n",
    "# ----- PYDCA calculate 'fields' ----- #\n",
    "\n",
    "# ------------------------------------ #\n",
    "\n",
    "# # -------------------------- PYDCA's direct info computation --------------------------------------- #\n",
    "num_unique_pairs = l * (l -1)                                                      \n",
    "num_unique_pairs /= 2                                                                            \n",
    "two_site_model_fields = np.zeros((np.int64(num_unique_pairs), 2, q), dtype=np.float64)           \n",
    "TOLERANCE = 1.0e-4                                                                            \n",
    "pair_counter = 0\n",
    "\n",
    "\n",
    "fig = plt.figure(figsize=(15,6))\n",
    "ax = fig.add_subplot(121)\n",
    "ax.set_title('w dist (full range)')\n",
    "ax.hist(w_2D.ravel(), alpha=0.5, bins=50, color= 'b', label='Tai weights')\n",
    "ax.hist(couplings.ravel(), alpha=0.5, bins=50,  color= 'r', label='PYDCA weights')\n",
    "ax = fig.add_subplot(122)\n",
    "ax.set_title('w dist (tele range)')\n",
    "ax.hist(w_2D.ravel(), alpha=0.5,range=(-2.5,5), bins=50, color= 'b', label='Tai weights')\n",
    "ax.hist(couplings.ravel(), alpha=0.5, bins=50, range=(-2.5,5),  color= 'r', label='PYDCA weights')\n",
    "ax.legend(loc='upper right', prop={'size':14})\n",
    "plt.savefig('w_val_distribution.pdf')\n",
    "# print('MF_di[0]:', MF_di[0, :10])\n",
    "# print('hybrid_di[0]:', hybrid_di[0, :10])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ER_di = np.delete(di, cols_removed,0)\n",
    "# ER_di = np.delete(ER_di, cols_removed,1)\n",
    "# Get best auc for PYDCA mf and plm (and plot the ROC curves)\n",
    "\n",
    "# which contact matrix do you want to use (full or curated)\n",
    "ct_mat = ct_full # full pdb contact matrix without columns removed\n",
    "ct_mat = ct  # full pdb contact matrix with conserved and bad columns removed (defined during pre-processing)\n",
    "plt.figure(figsize=(9.0,3.2))\n",
    "\n",
    "plt.title('ROC for PYDCA scores')\n",
    "\n",
    "# PYDCA-MF\n",
    "auc_pydca_MF = np.zeros(n)\n",
    "\n",
    "ax1 = plt.subplot2grid((1,2),(0,0))\n",
    "for i in range(n):\n",
    "    try:\n",
    "        p,tp,fp = roc_curve(ct_mat, pydca_MF_di, ct_thres[i])\n",
    "        if i in iplot:\n",
    "            ax1.plot(fp,tp,label='thres = %3.2f'%ct_thres[i])\n",
    "        auc_pydca_MF[i] = tp.sum()/tp.shape[0]\n",
    "    except(RuntimeWarning):\n",
    "        auc_MF[i] = 0\n",
    "ax1.set_xlim([0,1])\n",
    "ax1.set_ylim([0,1])\n",
    "ax1.set_xlabel('False Positive Rate')\n",
    "ax1.set_ylabel('True Positive Rate')\n",
    "ax1.set_title('PYDCA MF: %s' % method)\n",
    "        \n",
    "i0_pydca_MF = np.argmax(auc_pydca_MF)\n",
    "print('auc max:',ct_thres[i0_pydca_MF],auc_pydca_MF[i0_pydca_MF])\n",
    "p0_pydca_MF,tp0_pydca_MF,fp0_pydca_MF = tools.roc_curve(ct_mat,pydca_MF_di,ct_thres[i0_pydca_MF])\n",
    "\n",
    "\n",
    "# PYDCA-PLM\n",
    "auc_pydca_PLM = np.zeros(n)\n",
    "\n",
    "ax2 = plt.subplot2grid((1,2),(0,1))\n",
    "for i in range(n):\n",
    "    try:\n",
    "        p,tp,fp = roc_curve(ct_mat, pydca_PLM_di, ct_thres[i])\n",
    "        if i in iplot:\n",
    "            ax2.plot(fp,tp,label='thres = %3.2f'%ct_thres[i])\n",
    "        auc_pydca_PLM[i] = tp.sum()/tp.shape[0]\n",
    "    except(RuntimeWarning):\n",
    "        auc_pydca_PLM[i] = 0\n",
    "ax2.set_xlim([0,1])\n",
    "ax2.set_ylim([0,1])\n",
    "ax2.set_xlabel('False Positive Rate')\n",
    "ax2.set_ylabel('True Positive Rate')\n",
    "ax2.set_title('PYDCA PLM: %s' %method)\n",
    "        \n",
    "i0_pydca_PLM = np.argmax(auc_pydca_PLM)\n",
    "print('auc max:',ct_thres[i0_pydca_PLM],auc_pydca_PLM[i0_pydca_PLM])\n",
    "p0_pydca_PLM,tp0_pydca_PLM,fp0_pydca_PLM = tools.roc_curve(ct_mat,pydca_PLM_di,ct_thres[i0_pydca_PLM])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "pycharm": {
     "is_executing": true
    }
   },
   "outputs": [],
   "source": [
    "print('ER auc max:',ct_thres[i0_ER],auc_ER[i0_ER])\n",
    "print('DCA auc max:',ct_thres[i0_MF],auc_MF[i0_MF])\n",
    "\n",
    "print('PYDCA MF auc max:',ct_thres[i0_pydca_MF],auc_pydca_MF[i0_pydca_MF])\n",
    "print('PYDCA PLM auc max:',ct_thres[i0_pydca_PLM],auc_pydca_PLM[i0_pydca_PLM])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "pycharm": {
     "is_executing": true
    }
   },
   "outputs": [],
   "source": [
    "plt.figure(figsize=(12.0,3.2))\n",
    "\n",
    "# Plot ROC for optimal DCA vs optimal ER\n",
    "plt.subplot2grid((1,3),(0,0))\n",
    "plt.title('ROC at thres (MF, ER)\\n = (%3.2f, %3.2f)'%(ct_thres[i0_MF],ct_thres[i0_ER]))\n",
    "plt.plot(fp0_ER,tp0_ER,'b-',label=\"ER\")\n",
    "plt.plot(fp0_MF,tp0_MF,'r-',label=\"MF\")\n",
    "plt.plot(fp0_pydca_MF,tp0_pydca_MF,'ro',label=\"PYDCA MF\", markersize=1)\n",
    "plt.plot(fp0_pydca_PLM,tp0_pydca_PLM,'go',label=\"PYDCA PLM\", markersize=1)\n",
    "\n",
    "plt.plot([0,1],[0,1],'k--')\n",
    "plt.xlim([0,1])\n",
    "plt.ylim([0,1])\n",
    "plt.xlabel('False Positive Rate')\n",
    "plt.ylabel('True Positive Rate')\n",
    "plt.legend()\n",
    "\n",
    "# Plot AUC for DCA and ER\n",
    "plt.subplot2grid((1,3),(0,1))\n",
    "plt.title('AUC (MF, ER) max\\n = (%3.2f, %3.2f)' %(auc_MF[i0_MF], auc_ER[i0_ER]))\n",
    "plt.plot([ct_thres.min(),ct_thres.max()],[0.5,0.5],'k--')\n",
    "plt.plot(ct_thres,auc_ER,'b-',label=\"ER\")\n",
    "plt.plot(ct_thres,auc_MF,'r-',label=\"MF\")\n",
    "plt.plot(ct_thres,auc_pydca_MF,'ro',label=\"PYDCA MF\", markersize=1)\n",
    "plt.plot(ct_thres,auc_pydca_PLM,'go',label=\"PYDCA PLM\", markersize=1)\n",
    "plt.xlim([ct_thres.min(),ct_thres.max()])\n",
    "plt.ylim([0,max(auc_MF.max(), auc_ER.max())+0.1])\n",
    "plt.xlabel('distance threshold')\n",
    "plt.ylabel('AUC')\n",
    "plt.legend()\n",
    "\n",
    "# Plot Precision of optimal DCA and ER\n",
    "plt.subplot2grid((1,3),(0,2))\n",
    "plt.title('Precision at thres \\n(MF, ER) = (%3.2f, %3.2f)'%(ct_thres[i0_MF],ct_thres[i0_ER]))\n",
    "plt.plot( p0_ER,tp0_ER / (tp0_ER + fp0_ER),'b-',label='ER thres = %3.2f'%ct_thres[i0_ER])\n",
    "plt.plot( p0_MF,tp0_MF / (tp0_MF + fp0_MF),'r-',label='MF thres = %3.2f'%ct_thres[i0_MF])\n",
    "plt.plot( p0_pydca_MF,tp0_pydca_MF / (tp0_pydca_MF + fp0_pydca_MF),'ro',label='PYDCA MF %s thres = %3.2f' % (method, ct_thres[i0_pydca_MF]), markersize=1)\n",
    "plt.plot( p0_pydca_PLM,tp0_pydca_PLM / (tp0_pydca_PLM + fp0_pydca_PLM),'go',label='PYDCA PLM %s thres = %3.2f' % (method, ct_thres[i0_pydca_PLM]), markersize=1)\n",
    "\n",
    "plt.plot([0,1.],[.5,.5],'k--')\n",
    "plt.xlim([0,1])\n",
    "plt.ylim([0,1]) \n",
    "plt.xlabel('Recall (Sensitivity - P)')\n",
    "plt.ylabel('Precision (PPV)')\n",
    "plt.legend()\n",
    "\n",
    "plt.tight_layout(h_pad=1, w_pad=1.5)\n",
    "\n",
    "plt.savefig('ERvPYDCA_%s_comparison.pdf' % method)\n",
    "\n",
    "\n",
    "# tools.hide_toggle()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
